src/runtime-rs/crates/hypervisor/src/device/topology.rs

```diff
@@ -0,0 +1,293 @@
+    pub resource_reserved: ResourceReserved,
+
+    // romfile specifies the ROM file being used for this device.
+    pub romfile: String,
+}
+
+// PCIe Root Complex
+#[derive(Clone, Debug, Default)]
+pub struct PCIeRootComplex {
+    pub root_bus: String,
+    pub root_bus_address: String,
+    pub root_bus_devices: HashMap<String, PCIeEndpoint>,
+}
+
+#[derive(Debug, Default)]
+pub struct PCIeTopology {
+    pub hypervisor_name: String,
+    pub root_complex: PCIeRootComplex,
+
+    pub bridges: u32,
+    pub pcie_root_ports: u32,
+    pub hotplug_vfio_on_root_bus: bool,
+}
+
+impl PCIeTopology {
+    // As some special case doesn't support PCIe devices, there's no need to build a PCIe Topology.
+    pub fn new(config_info: Option<&TopologyConfigInfo>) -> Option<Self> {
+        // if config_info is None, it will return None.
+        let topo_config = config_info?;
+
+        let root_complex = PCIeRootComplex {
+            root_bus: DEFAULT_PCIE_ROOT_BUS.to_owned(),
+            root_bus_address: DEFAULT_PCIE_ROOT_BUS_ADDRESS.to_owned(),
+            root_bus_devices: HashMap::with_capacity(PCIE_ROOT_BUS_SLOTS_CAPACITY as usize),
+        };
+
+        Some(Self {
+            hypervisor_name: topo_config.hypervisor_name.to_owned(),
+            root_complex,
+            bridges: topo_config.device_info.default_bridges,
+            pcie_root_ports: topo_config.device_info.pcie_root_port,
+            hotplug_vfio_on_root_bus: topo_config.device_info.hotplug_vfio_on_root_bus,
+        })
+    }
+
+    pub fn insert_device(&mut self, ep: &mut PCIeEndpoint) -> Option<PciPath> {
+        let to_pcipath = |v: u32| -> PciPath {
+            PciPath {
+                slots: vec![PciSlot(v as u8)],
+            }
+        };
+
+        let to_string = |v: u32| -> String { to_pcipath(v).to_string() };
+
+        // find the first available index as the allocated slot.
+        let allocated_slot = (0..PCIE_ROOT_BUS_SLOTS_CAPACITY).find(|&i| {
+            !self
+                .root_complex
+                .root_bus_devices
+                .contains_key(&to_string(i))
+        })?;
+
+        let pcipath = to_string(allocated_slot);
+
+        // update pci_path in Endpoint
+        ep.pci_path = to_pcipath(allocated_slot);
+        // convert the allocated slot to pci path and then insert it with ep
+        self.root_complex
+            .root_bus_devices
+            .insert(pcipath, ep.clone());
+
+        Some(to_pcipath(allocated_slot))
+    }
+
+    pub fn remove_device(&mut self, device_id: &str) -> Option<String> {
+        let mut target_device: Option<String> = None;
+        self.root_complex.root_bus_devices.retain(|k, v| {
+            if v.device_id() != device_id {
+                true
+            } else {
+                target_device = Some((*k).to_string());
+                false
+            }
+        });
+
+        target_device
+    }
+
+    pub fn update_device(&mut self, ep: &PCIeEndpoint) -> Option<PciPath> {
```

Comment:
update_device here means that device_id remains the same but pci_addr needs to be changed?

Reply:
comments added to explain it.